src/zkstar.rs

```diff
@@ -0,0 +1,160 @@
+use crate::errors::InternalError;
+use crate::utils;
+use libpaillier::unknown_order::BigNumber;
+use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+
+/// Tool for verifying and operating on elements of the multiplicative
+/// group of integers mod `N`, for some modulus `N`.
+///
+/// The builder defines the modulus `N` for all [`ZStarN`] that it creates.
+/// This simplifies the `ZStarN` type and allows it to be serialized _without_
+/// duplicating the modulus.
+#[derive(Serialize)]
+struct ZStarNBuilder {
+    modulus: BigNumber,
+}
+#[allow(unused)]
+impl ZStarNBuilder {
+    pub fn new(n: BigNumber) -> Self {
+        ZStarNBuilder { modulus: n }
+    }
+    /// One of two ways of constructing an element of [ZStarN] by validating that an unverifier instance of [ZStarNUnverified]
+    /// is properly constructed with respect to the current [ZStarNBuilder]
+    /// (the other way is by randomly sampling the element in the multiplicative group modulo n)
+    pub fn validate(&self, unverified: ZStarNUnverified) -> Result<ZStarN, InternalError> {
+        if unverified.value().is_zero() {
+            return Err(InternalError::IsZero);
+        } else if unverified.value() > self.modulus() {
+            return Err(InternalError::LargerThanModulus);
+        } else if unverified.value() < &BigNumber::zero() {
+            return Err(InternalError::NegativeElement);
+        }
+        let result = unverified.value().gcd(self.modulus());
+        if result != BigNumber::one() {
+            return Err(InternalError::NotCoprime);
+        }
+        Ok(ZStarN {
+            value: unverified.value().to_owned(),
+            builder: self,
+        })
+    }
+    /// Return the modulus of [ZStarNBuilder]
+    fn modulus(&self) -> &BigNumber {
+        &self.modulus
+    }
+}
+
+/// A multiplicative group of elements modulo `N` as defined by [ZStarNBuilder]
+/// Elements of this group are in the interval (0, N) and validated to be as such.
```

Comment:
nits: 
- "An element in the multiplicative group..."
- code formatting on link
- `.` at the end of the first sentence
- add a newline between the two sentences (so when a text editor hovers on a `ZStarN`, it just pops up the first sentence)